### ISSUE TITLE
Uncaught TypeError with PressGrid theme

### ISSUE THREAD

**Issue by F4bri92** [comfusion/hyperdrive#80]
Trying hyperdrive I get in console

> Uncaught TypeError: Cannot read property 'extend' of undefined
>     at <anonymous>:27:20271
>     at <anonymous>:27:21198
>     at e ((index):11)
>     at Object.then ((index):11)

plus other similar errors and then everything javascript-related doesn't work.
This is <anonymous>:27:20271-21198

`.extend(mejs.MepDefaults,{playText:mejs.i18n.t("Play"),pauseText:mejs.i18n.t("Pause")}),a.extend(MediaElementPlayer.prototype,{buildplaypause:function(b,c,d,e){function f(a){"play"===a?(i.removeClass("mejs-play").addClass("mejs-pause"),j.attr({title:h.pauseText,"aria-label":h.pauseText})):(i.removeClass("mejs-pause").addClass("mejs-play"),j.attr({title:h.playText,"aria-label":h.playText}))}var g=this,h=g.options,i=a('<div class="mejs-button mejs-playpause-button mejs-play" ><button type="button" aria-controls="'+g.id+'" title="'+h.playText+'" aria-label="'+h.playText+'"></button></div>').appendTo(c).click(function(a){return a.preventDefault(),e.paused?e.play():e.pause(),!1}),j=i.find("button");f("pse"),e.addEventListener("play",function(){f("play")},!1),e.addEventListener("playing",function(){f("play")},!1),e.addEventListener("pause",function(){f("pse")},!1),e.addEventListener("paused",function(){f("pse")},!1)}})}`

This is (index):11
`!function(e,t){"object"==typeof exports&&"undefined"!=typeof module?module.exports=t():"function"==typeof define&&define.amd?define(t):e.fetchInject=t()}(this,function(){"use strict";const e=function(e,t,n,o,r,i,c){i=t.createElement(n),c=t.getElementsByTagName(n)[0],i.appendChild(t.createTextNode(o.text)),i.onload=r(o),c?c.parentNode.insertBefore(i,c):t.head.appendChild(i)};return function(t,n){if(!t||!Array.isArray(t))return Promise.reject(new Error("`inputs` must be an array"));if(n&&!(n instanceof Promise))return Promise.reject(new Error("`promise` must be a promise"));const o=[],r=n?[].concat(n):[],i=[];return t.forEach(e=>r.push(window.fetch(e).then(e=>[e.clone().text(),e.blob()]).then(e=>Promise.all(e).then(e=>{o.push({text:e[0],blob:e[1]})})))),Promise.all(r).then(()=>{o.forEach(t=>{i.push({then:n=>{t.blob.type.includes("text/css")?e(window,document,"style",t,n):e(window,document,"script",t,n)}})});return Promise.all(i)})}});`

Other infos:
Wp 4.8, Theme: Press-Grid

Thanks for any help

**Comment by jhabdas** [comment]
Thanks for the info @F4bri92. Happy to take a look. Could you point me to the theme and let me know which beta version your running?

**Comment by F4bri92** [comment]
I'm using beta6, theme is this one: https://themeforest.net/item/pressgrid-frontend-publish-reaction-multimedia-theme/18701111 Can I send you the theme via email?

**Comment by jhabdas** [comment]
@F4bri92 fire away. josh@habd.as

**Comment by F4bri92** [comment]
@jhabdas  Just sent :)

**Comment by jhabdas** [comment]
Payload beamed successfully and sent to engineering for investigation.

**Comment by F4bri92** [comment]
@jhabdas Thank you!

**Comment by jhabdas** [comment]
_Bridge to engineering. What's the status on PressGrid? Psstssh_

...

_Engineering to Bridge. It appears PressGrid is loading Media Element scripts before jQuery, which they require, causing a failure in the Media Element extension sequence when the Hyperdrive is engaged, causing what appears to be a malformed nebula. We're sending a dump of the destination coordinates calculated during the thruster calibration sequence for you to have a look. Over._

```json
[["wp-mediaelement","/wp-includes/js/mediaelement/wp-mediaelement.min.js",[["mediaelement","/wp-includes/js/mediaelement/mediaelement-and-player.min.js?ver=2.22.0",[]],["mediaelement","",[["jquery","",[["jquery-core","/wp-includes/js/jquery/jquery.js?ver=1.12.4",[]]]]]]]],["press-grid-lib-packages","http://wordpress.dev/wp-content/themes/press-grid/js/packages.min.js",[["jquery","",[["jquery-core","/wp-includes/js/jquery/jquery.js?ver=1.12.4",[]]]]]],["press-grid-scripts","http://wordpress.dev/wp-content/themes/press-grid/js/scripts.min.js",[["jquery","",[["jquery-core","/wp-includes/js/jquery/jquery.js?ver=1.12.4",[]]]]]],["press-grid-app-script","http://wordpress.dev/wp-content/themes/press-grid/js/app.js",[["jquery","",[["jquery-core","/wp-includes/js/jquery/jquery.js?ver=1.12.4",[]]]],["mediaelement","/wp-includes/js/mediaelement/mediaelement-and-player.min.js?ver=2.22.0",[]],["mediaelement","",[["jquery","",[["jquery-core","/wp-includes/js/jquery/jquery.js?ver=1.12.4",[]]]]]],["wp-playlist","/wp-includes/js/mediaelement/wp-playlist.min.js",[]],["wp-playlist","",[["wp-util","/wp-includes/js/wp-util.min.js",[]],["wp-util","",[["underscore","/wp-includes/js/underscore.min.js?ver=1.8.3",[]],["jquery","",[["jquery-core","/wp-includes/js/jquery/jquery.js?ver=1.12.4",[]]]]]],["backbone","/wp-includes/js/backbone.min.js?ver=1.2.3",[]],["backbone","",[["underscore","/wp-includes/js/underscore.min.js?ver=1.8.3",[]],["jquery","",[["jquery-core","/wp-includes/js/jquery/jquery.js?ver=1.12.4",[]]]]]],["mediaelement","/wp-includes/js/mediaelement/mediaelement-and-player.min.js?ver=2.22.0",[]],["mediaelement","",[["jquery","",[["jquery-core","/wp-includes/js/jquery/jquery.js?ver=1.12.4",[]]]]]]]]]]]
```

_Bridge: Very good. We'll have our droid run some diagnostics and compare it with known anomalies based on a cross-reference of database records within the hour._

**Comment by jhabdas** [comment]
_Bridge reporting. Engineering, come in please. Our droid was able to determine PressGrid was improperly loading the Media Element scripts. We're sending some diagnostics showing the two lines of code we change for PressGrid to fix the Media Element extends issue..._

**dumping diagnostics:**
```php
wp_enqueue_script( 'wp-mediaelement' ); // incorrect usage
```
```php
wp_enqueue_script('press-grid-app-script', get_template_directory_uri() . '/js/app.js', array('jquery', 'wp-mediaelement', 'wp-playlist'), false, true ); // proper usage
```

_Engineering. We're also going to need investigation of another issue in PressGrid. It appears resolving the jQuery extension sequencing issue the malformed nebula has disappeared, and appearing in it's place is a dime-sized black hole called React which is now making it difficult to see the card stacks. Please investigate and report back. Over. Psssdfshfsdf._

...

_Engineering to Bridge. We read you loud and clear and will begin the investigation into reason for the black hole immediately. Over and out._

**Comment by jhabdas** [comment]
_Engineering to bridge. It seems we have gone directly into the black hole and emerged on the other side. While passing through the event horizon we noticed the following..._

...

We made a few adjustments to the Hyperdrive beta 6 in `master`. PressGrid can load now without error and is approximately **18% faster over 3G** connections. We also observed the above diagnostics are no longer necessary with the latest updates. 

But there are some caveats:

- Theme needs to define script dependencies as per guidance in #70 for it to work with Hyperdrive. Specifically the `press-grid-app-script` script depends on `press-grid-scripts` but dependency has not been established.
- Upon establishing this dependency, all scripts can load without error on fast connections, but there’s still a race condition in the implementation as it needs standards-based HTML5 [`readyState`](http://devdocs.io/dom/document/readystate) handlers as suggested in #70.
- We also observed the app has two entry points to run its client-side scripts. Which means they may need a more streamlined way of initiating execution to avoid potential warp failure.

...

_Bridge. We're attaching a screenshot showing scripts fully functional and loading without error in case the SendGrid maintainer would like to add support for Hyperdrive and needs to see it loading without error to reduce any uncertainties. Over._

...

_Bridge to Engineering. Great work. We became aware the recent adjustments made brought unit test coverage up to 91%. That's outstanding work. Let us carry on with our mission and perhaps revisit SendGrid at some point in a possible future. Pssssshhtttt._

<img width="550" alt="screen shot 2017-06-14 at 11 03 55 am" src="https://user-images.githubusercontent.com/440298/27115996-fb131362-5100-11e7-9d84-1f641332c3b7.png">

@F4bri92 Please feel free to try out the latest master. I'll be cutting beta 7 soon as a result of this investigation and some important adjustments made during it. If there's anything I can do to help with the theme adjustments needed to enable Hyperdrive support in SendGrid please let me know. Godspeed.

![screen shot 2017-06-14 at 12 52 48 pm](https://user-images.githubusercontent.com/440298/27116029-3ad99cfa-5101-11e7-8c9d-a89dadc67c40.png)

**Comment by F4bri92** [comment]
Ok, everything seems to be ok except the "reaction" buttons in posts.

```
Uncaught ReferenceError: theme_options is not defined
    at HTMLAnchorElement.<anonymous> (<anonymous>:5:28531)
    at HTMLAnchorElement.dispatch (<anonymous>:1:44454)
    at HTMLAnchorElement.r.handle (<anonymous>:1:41183)
```

**Comment by jhabdas** [comment]
Thanks for checking, @F4bri92. Looking at the theme's code I could tell the author is a solid coder. With the info above they should be able to take the theme the last mile for full support. Perhaps they would be interested in a support request. I can promise they've never seen a plugin like this so they may be very interested.

**Comment by F4bri92** [comment]
Could hyperdrive be breaking _wp_add_inline_script_?

theme_options is part of inline script added trough:
`wp_add_inline_script('press-grid-lib-packages', Press_Grid_Tpl::inline_script() );`

**Comment by jhabdas** [comment]
Oh interesting. I saw that but am not sure exactly what the theme author is trying to achieve there. Maybe it's possible to create a child theme to add a compatibility layer. I'm not sure. Please see #70 for plugin and theme author compatibility guidance. It's still a work in progress, and I may need to look into inline scripts more, so thanks for bringing it up. Good stuff.

**Comment by F4bri92** [comment]
@jhabdas Thank you for all the help! You rock! 🥇 

That function take some infos from cookies and current columns and prints something like:
```
var theme_options = { ajax_url: "https://wp.url/wp-admin/admin-ajax.php" };
var themeton_reaction_of_posts = [];
var themeton_content_col = "col-xl-3 col-lg-4 col-md-4 col-sm-6 col-xs-12";
```

I temporarily fixed this with `echo "<script>".Press_Grid_Tpl::inline_script()."</script>"` in header.php

**Comment by jhabdas** [comment]
Hell yeah. Does React work now?

**Comment by F4bri92** [comment]
Yep! 👍

**Comment by jhabdas** [comment]
w00t! All right let's see about triaging any remaining issues. Would you be so kind as to throttle your network connection to 3G  from dev tools disable your browser cash and try reloading the page to see what happens?

**Comment by F4bri92** [comment]
![schermata a 2017-06-14 10-31-19](https://user-images.githubusercontent.com/8255936/27122975-a920cad6-50ec-11e7-8c0c-65bdc64456ad.png)

**Comment by jhabdas** [comment]
That's so fetch! ;)

**Comment by jhabdas** [comment]
Opened #82 based on your findings @F4bri92, provided some investigation and documented your workaround. I'm curious if creating a child theme as a kind of [_ponyfill_](https://github.com/sindresorhus/ponyfill), would be a reasonable step towards providing a workaround without requiring individuals to hack theme internals.

**Comment by jhabdas** [comment]
@F4bri92 I noticed the waterfall was getting blocked by styles, including an external font request and potentially causing a single-point of failure. As a result you may want to check your waterfall again with `1.0.0-beta.8`, just released. [CHANGELOG](https://github.com/comfusion/hyperdrive/blob/master/CHANGELOG.md#100-beta8-2017-06-15)

**Comment by F4bri92** [comment]
Fetch works but resources loaded through css like webfonts don't load (err 404).
Missing /wp-content/themes/theme-name/ in url...

**Comment by jhabdas** [comment]
Thanks for the quick response. Will investigate. Haven't seen that yet with web fonts.

**Comment by jhabdas** [comment]
Ah, I know what it is. I have seen it, with PhotoSwipe. CSS may assume a relative path which has changed. Given that I may surface an option to give admins more control. Will mull on it as CSS may be updated as well based on my work with PhotoSwipe (which is why I created Fetch Injection).

**Comment by jhabdas** [comment]
@F4bri92 I pushed up a new beta removing the async style loading. While I want to introduce as much change as possible as early as possible, I made a commitment not to add an admin menu in 1.0 and, as it stands, the async style loading can't go out without one so I've pulled it for now. Thanks again for testing.

**Comment by jhabdas** [comment]
Closing this for now to keep noise down under issues for those approaching Hyperdrive as this is very theme specific.
